kernel/fs/vfs.c

```diff
@@ -804,13 +840,80 @@ fs_node_t *kopen(char *filename, uint32_t flags) {
 		/* Search the active directory for the requested directory */
 		debug_print(INFO, "... Searching for %s", path_offset);
 		node_next = finddir_fs(node_ptr, path_offset);
-		free(node_ptr);
+		close_fs(node_ptr);
 		node_ptr = node_next;
 		if (!node_ptr) {
 			/* We failed to find the requested directory */
 			free((void *)path);
 			return NULL;
-		} else if (depth == path_depth - 1) {
+		}
+		/* 
+		 * This test is a little complicated, but we basically always resolve symlinks in the
+		 * of a path (like /home/symlink/file) even if O_NOFOLLOW and O_PATH are set. If we are
+		 * on the leaf of the path then we will look at those flags and act accordingly
+		 */
+		if ((node_ptr->flags & FS_SYMLINK) &&
+				!((flags & O_NOFOLLOW) && (flags & O_PATH) && depth == path_depth - 1)) {
+			/* This ensures we don't return a path when NOFOLLOW is requested but PATH
+			 * isn't passed.
+			 */
+			debug_print(NOTICE, "resolving symlink at %s", node_ptr->name);
+			if ((flags & O_NOFOLLOW) && depth == path_depth - 1) {
+				/* TODO(gerow): should probably be setting errno from this */
+				debug_print(NOTICE, "Refusing to follow final entry for open with O_NOFOLLOW for %s.", node_ptr->name);
+				free((void *)path);
+				close_fs(node_ptr);
+				return NULL;
+			}
+			if (symlink_depth >= MAX_SYMLINK_DEPTH) {
+				/* TODO(gerow): should probably be setting errno from this */
+				debug_print(WARNING, "Reached max symlink depth on %s.", node_ptr->name);
+				free((void *)path);
+				close_fs(node_ptr);
+				return NULL;
+			}
+			/* 
+			 * This may actually be big enough that we wouldn't want to allocate it on
+			 * the stack, especially considering this function is called recursively
+			 */
+			char symlink_buf[MAX_SYMLINK_SIZE];
+			int len = node_ptr->readlink(node_ptr, symlink_buf, sizeof(symlink_buf));
+			if (len < 0) {
+				/* TODO(gerow): should probably be setting errno from this */
+				debug_print(WARNING, "Got error %d from symlink for %s.", len, node_ptr->name);
+				free((void *)path);
+				close_fs(node_ptr);
+				return NULL;
+			}
+			if (symlink_buf[len - 1] != '\0') {
```

Comment:
Yup. From what I gathered readlink is supposed to null-terminate the buffer, so I just assumed ext2 null terminated the symlink on-disk. That's probably not actually the case...